ROADMAP.md

```diff
@@ -0,0 +1,579 @@
+
+### 🗣️ Polite peering and wire protocol
+
+**What?** Peers don't behave well with one another. They do not send
+DISCONNECT messages reporting the reason for disconnection, they do not
+warn when they're about to close a connection, they don't give peers a
+chance to keep important connections open, etc.
+
+Furthermore, IDENTIFY is an element that other parts of the system
+depend heavily on (e.g. AutoRelay, DHT, etc.), yet it does not have a
+special status. Similarly, PING is a housekeeping protocol.
+
+Grouping all these essential messages into a solid *Wire protocol* that
+covers connection lifecycle, heartbeats, health, status, diagnostics,
+identity, etc. may simplify things. An open question is where
+Multistream would fit into this model: could we conflate it into the
+Wire protocol?
+
+**Why?** Peers act haphazardly when observed from the outside, as they
+do not have the means to act more collaboratively. We are lacking a wire
+protocol that governs the connection between two peers. Negotiating lots
+of piecemeal protocols is inefficient, especially for essential stuff.
+
+**Links:**
+
+-   [Discuss: batch vs. on-demand connection
+    pruning](https://github.com/libp2p/go-libp2p-connmgr/issues/19).
+
+
+### ☎️ Reducing the dial fail rate
+
+**What?** Reducing the dial fail rate implies improvements in several
+areas.
+
+-  **Outbound dials:** We should strategise dials based on context, or based on
+   a heuristic determined by the application. For example, apps designed for
+   local exchange of data will naturally prefer to try local/physical endpoints
+   over public ones first. We may want to dial relay addresses last, dial newer
+   addresses first, etc.
+
+   - Users should also have the capability to customise how addresses
+     are resolved, how dials are throttled, how connections are
+     selected if multiple dials succeeded, etc. In a nutshell, the
+     dialing behaviour should be fully customisable per host.
+
+-  **Context-sensitive advertisement of self-addresses:** private
+   addresses should not leak into public DHTs, both for connectivity
+   and security reasons (cf. IPFS DHT pollution in 2016). Similarly,
+   local discovery mechanisms should only transmit private addresses.
+   The DHT topology question becomes relevant to define scopes (e.g.
+   federated/concentric/shared DHTs for different scopes). Also, when
+   dealing with multi-layer NATs, there may be better routes between
+   peers behind regional or ISP NATs other than their public
+   endpoints.
+
+- **Connectivity-related problems:** see [Unprecedented global
+  connectivity](#unprecedented-global-connectivity) and
+  [Improve NAT traversal](#improve-nat-traversal).
+
+**Why?** Dialing is a crucial function of our system and using a
+non-deterministic, scattershot approach is inefficient.
+
+**Links:**
+
+- [Reliable dialing](https://github.com/libp2p/libp2p/issues/27).
+
+### 🔀 Peer exchange protocol
```

Comment:
Maybe move this one up since it's actively being worked on?

Reply:
The mDNS part, yes.
Furthermore, I believe we're already running two DHTs in parallel, one for the public internet and one for local networks.
@mxinden Is it possible this issue is already resolved (assuming we can land my mdns changes), and we can remove this section entirely?

Reply:
This goes beyond mDNS and beyond two (local and global) DHTs.

See https://github.com/libp2p/notes/issues/3 for local network peer exchange and https://github.com/libp2p/notes/issues/7 and https://github.com/libp2p/specs/issues/222 on a peer exchange protocol in general.

Also related is the [rendezvous protocol](https://github.com/libp2p/specs/blob/master/rendezvous/README.md) solving a similar problem with a smaller number of (rendezvous) nodes instead of a fully decentralized peer exchange protocol.

> Is it possible this issue is already resolved (assuming we can land my mdns changes), and we can remove this section entirely?

I don't think this is resolved by providing mDNS and two DHTs. That said (a) I don't feel strongly about it and (b) I am not sure it requires a whole section in our roadmap on its own. What do you all think? Should we remove it? Should we move it up?

Reply:
Marking this conversation as resolved. Feel free to un-resolve in case you want to continue the stalled discussion.